aten/src/ATen/native/TensorIterator.cpp

```diff
@@ -98,18 +103,47 @@ compute_result_type(at::ArrayRef<OperandInfo> operands,
 
 std::tuple<Device, ScalarType> TensorIterator::compute_common_type() {
   // See [Result type computation] in TensorIterator.h
+
   auto result_type =
       compute_result_type(operands_,
-        [](const Tensor& t) { return t.dim() > 0; },
-        [](const Tensor& t) { return !t.unsafeGetTensorImpl()->is_wrapped_number(); },
-        [](const Tensor& t) { return true; });
+        [](const OperandInfo& op) { return op.tensor.dim() > 0; },
+        [](const OperandInfo& op) { return !op.tensor.unsafeGetTensorImpl()->is_wrapped_number(); },
+        [](const OperandInfo& op) { return true; });
+
+  // if non-zero-dim tensor result is an integral type and there's a zero-dim
+  // floating point operand, we'll promote the the floating point type.
+  if (isIntegralType(std::get<1>(result_type))) {
+    auto alternate = compute_result_type(operands_,
+        [](const OperandInfo& op) {
+          return isFloatingType(op.tensor.scalar_type()) && op.tensor.dim() == 0;
+        }
+    );
+    if (std::get<1>(alternate) != ScalarType::Undefined) {
+      // preserve device from original result
+      return std::make_tuple(std::get<0>(result_type), std::get<1>(alternate));
+    }
+  }
 
   TORCH_INTERNAL_ASSERT(std::get<1>(result_type) != ScalarType::Undefined);
   return result_type;
 }
 
+// just checks for safe promotion.
+static bool can_promote(const ScalarType& from, const ScalarType& to) {
+  // we disallow float -> integral, e.g., float_tensor *= int is disallowed.
+  if (isFloatingType(from) && isIntegralType(to)) {
+    return false;
+  }
+  // if destination dtype is smaller, would lose information.
+  if (elementSize(from) > elementSize(to)) {
```

Comment:
This is different from NumPy's behavior. The default casting in NumPy for these sorts of ops is "same_kind", which does not have this elementSize restriction.

In particular, this is allowed by NumPy: `float_tensor *= double_tensor`

However, NumPy does not allow `uint8_tensor *= int16_tensor` because "unsigned" and "signed" are different kinds.

Reply:
updated to handle signed and not to block downcast on size.